Improve performance of WebUtility.Decode for non-escaped strings

hughbe opened the pull request:
- Cuts allocations down to 1/3 of the original
- Doubles performance (in time)

- We still allocate a little bit, as removing all allocations would harm the performance of strings that actually need escaping

## Benchmark - escaping needed
- No performance regressions

![benchmark escaping](https://cloud.githubusercontent.com/assets/1275900/14454716/c85fc7bc-0093-11e6-8c65-3eba5ce34729.png)


## Benchmark - no escaping needed
![benchmark no escaping](https://cloud.githubusercontent.com/assets/1275900/14454713/c5124e54-0093-11e6-983b-ad3ae9e654ff.png)

## Benchmark Code
<details>
<summary>Click here</summary>
```
static void Main(string[] args)
{
	// Escaping
    TimeAction("Old: ", () => Old.UrlDecode("%ABabc"));
    TimeAction("New: ", () => New.UrlDecode("%ABabc"));

    // No escaping
    TimeAction("Old: ", () => Old.UrlDecode("abc"));
    TimeAction("New: ", () => New.UrlDecode("abc"));

    Console.ReadLine();
}

public static void TimeAction(string prefix, Action action)
{
    var sw = new Stopwatch();
    for (int iter = 0; iter < 5; iter++)
    {
        int gen0 = GC.CollectionCount(0);
        sw.Restart();
        for (int i = 0; i < 10000000; i++)
        {
            action();
        }
        sw.Stop();
        Console.WriteLine($"{prefix}Time: {sw.Elapsed.TotalSeconds}\tGC0: {GC.CollectionCount(0) - gen0}");
    }
}
```
</details>

/cc @stephentoub @jamesqo @davidsh 
Fixes #6542 together with #7546

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -610,6 +612,9 @@ private static bool StringRequiresHtmlDecoding(string s)
         private struct UrlDecoder
         {
+            public bool _containsUnsafe;
+            public bool _containsSpaces;
```
Why store these as fields on the UrlDecoder?  Seems like they could just be a single local ```needsDecoding```, and then instead of doing ```return helper.GetString();```, you'd do ```return needsDecoding ? helper.GetString() : value;```.

hughbe replied:
I agree with moving them out of UrlDecoder.
I would like to keep the fields `containsUnsafe` and `containsSpaces` as separate variables, as I'm working on a PR that's gonna heavily optimize `UrlEncode` and `UrlDecode` for strings that only need space encoding/decoding. Is that OK?

stephentoub replied:
I'd prefer it if you kept them as a single local now.  If a subsequent change needs them separated, then that change can do it.  Otherwise, if for example that change never materializes or gets merged, we're left with debt.

hughbe replied:
Sure thing!

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -425,7 +427,7 @@ private static string UrlDecodeInternal(string value, Encoding encoding)
                     helper.AddChar(ch);
```
I don't understand how this AddByte vs AddChar mechanism works.  It looks like GetString dumps all of the bytes to the end of the chars... doesn't that result in bytes that came in the middle of the input getting decoded out of order?

stephentoub replied:
Oh, nevermind, I see, each call to AddChar dumps any bytes that exist.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -626,17 +631,23 @@ private struct UrlDecoder
 
-                _charBuffer = new char[bufferSize];
-                
+                _charBuffer = null; // char buffer created on demand
```
Is this really useful?  Empty string is already special cased to avoid decoding.  So isn't this only valuable in the case where all of the inputs are bytes rather than chars, and to enable that you're then doing an extra null check branch on every AddChar?  Doesn't seem like a good tradeoff.

hughbe replied:
Well I found that this halved allocations for the `UrlDecode("noescaping")` case, as it seems that ASCII chars are added to the decoder as `AddByte` not as `AddChar`

hughbe replied:
Just a follow up: its due to [this](https://github.com/hughbe/corefx/blob/webutility-decode/src/System.Runtime.Extensions/src/System/Net/WebUtility.cs#L424) line of code.

I don't really understand why "// 7 bit have to go as bytes because of Unicode" but hey

stephentoub replied:
I see, ok.  (Seems strange then that we're lazily allocating the byte[] array; doing so suggests we don't expect any ASCII chars in common input, but let's not change it.)

hughbe replied:
understood - obviously I'm not planning to change any of that behaviour i this PR, but do you have a theory as to why we add ascii chars as bytes not chars?

stephentoub replied:
I don't.  Maybe @ellismg or @tarekgh would know?

tarekgh replied:
if you look at UrlDecoder, it keeps character and byte arrays, ascii encoded characters will kept as bytes and the rest kept as characters. when calling UrlDecoder.GetString it'll convert the bytes back to characters. maybe i tis done that way for size optimization as most the Urls today contains ascii characters.

stephentoub replied:
> maybe it is done that way for size optimization as most the Urls today contains ascii characters.

But the current code is always allocating the full length char array anyway.

tarekgh replied:
right but they allocate the array only when needed. if you encounter asciii Urls most of the time, then will not need to allocate for characters array till you flush the decoder. just to note that what I am seeing but I admit I don't really know the actual intention of who wrote the code :-)

stephentoub replied:
> right but they allocate the array only when needed

No. In the code before this PR, the char[] is always allocated (unless the string is empty).

tarekgh replied:
right. looking more, it looks when having ascii characters, it needs to be encoded by the encoder. bu tif the character is not ascii will be stored as it is without encoding. so I believe what they are doing is collecting ascii charcaters in the array, and everytime need to flush it, will run the encoder just once on the array before the flush. so it is just optimizing to not calling the encoder with every ascii character.

hughbe replied:
> so it is just optimizing to not calling the encoder with every ascii character.

but doesn't the implementation check if there are any bytes in the buffer, and if there are none in the buffer it doesn't flush any bytes

tarekgh replied:
UrlDecoder is collecting the ascii bytes till either AddChar or GetString get called. at that time it will flush the bytes after encoding it. and then will start collecting the new encountered bytes again and so on. 
imagine you have string like aaaaaNaaaaNaaaaNaaa
where a is ascii character and N is not ascii, this will have the byte array get filled 4 times and encoded 4 times. if you didn't collect the ascii in byte array as the code is doing, this mean you'll need to call the encoder the number of 'a' characters in the string.

hughbe replied:
understood, thanks

jamesqo commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -425,6 +428,11 @@ private static string UrlDecodeInternal(string value, Encoding encoding)
 
+            if (!needsDecoding)
+            {
+                // No decoding needd
```
Typo: should be 'needed'

hughbe commented:
Test Innerloop Ubuntu Debug Build and Test

hughbe commented:
Test Innerloop CentOS7.1 Release Build and Test
Test Innerloop Windows_NT Debug Build and Test

stephentoub commented:
@dotnet-bot test code coverage please

hughbe commented:
@stephentoub  is this new!

stephentoub commented:
> is this new

nope

> can non-Microsoft people use that code coverage feature

yup

hughbe commented:
> nope

That's cool, I've never noticed that before. Is the idea that it fails if there is an overall decrease in coverage or does it just publish a report? (I guess we'll find out!)

stephentoub commented:
> Is the idea that it fails if there is an overall decrease in coverage or does it just publish a report? (I guess we'll find out!)

It doesn't perform any comparison, but it archives the report so you can click through to view it.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -415,6 +417,7 @@ private static string UrlDecodeInternal(string value, Encoding encoding)
 
                         // don't add as char
                         helper.AddByte(b);
+                        needsDecoding = true;
```
According to the code coverage report, there are missed branches for this if.  Can you please add tests to ensure it's fully covered?

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -425,6 +428,11 @@ private static string UrlDecodeInternal(string value, Encoding encoding)
                     helper.AddChar(ch);
             }
```
According to the code coverage report, this helper.AddChar(ch) is never executed by tests, which means some of the code paths you added around lazily initializing the array are never touched.  Can you please add tests cases to ensure everything is covered appropriately?

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -425,6 +428,11 @@ private static string UrlDecodeInternal(string value, Encoding encoding)
+            {
+                // No decoding needed
+                return value;
+            }
```
If it were me, I'd have written this like:
```C#
return needsDecoding ? helper.GetString() : value;
```

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -626,6 +634,9 @@ private struct UrlDecoder
             private void FlushBytes()
             {
                 Debug.Assert(_numBytes > 0);
+                if (_charBuffer == null)
```
There are no tests covering the case where _charBuffer is already initialized.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -647,6 +658,9 @@ internal void AddChar(char ch)
                 if (_numBytes > 0)
                     FlushBytes();
 
+                if (_charBuffer == null)
```
There no tests executing this AddChar method.

hughbe commented:
Thanks, I agree its a good idea to add some more tests to cover all the branches I've changed. I also added some more tests for invalid percent encoding, lowercase hex, space escaping and non-ASCII chars.
This brings coverage up to 100% for UrlDecodeInternal and almost 100% for UrlDecoder - I think the missing coverage may be dead code (line 683)

stephentoub commented:
LGTM

@davidsh?

davidsh commented:
LGTM.

davidsh commented:
CI Failures:
http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/windows_nt_debug_prtest/2/

(SslStream)

davidsh commented:
Test Innerloop Windows_NT Debug Build and Test

davidsh commented:
@dotnet-bot Test Innerloop CentOS7.1 Release Build and Test

hughbe commented:
Glad to see this merged. Should I send over a similar PR to coreclr, or will this be automatically handled by the netfx-port-consider label
